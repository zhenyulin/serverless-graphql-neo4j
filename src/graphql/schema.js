import { makeExecutableSchema } from 'apollo-server-lambda';
import { augmentSchema } from 'neo4j-graphql-js';

import typeDefs from './type';
import resolvers from './resolver';

export default augmentSchema(
	makeExecutableSchema({
		typeDefs,
		resolvers,
		resolverValidationOptions: {
			requireResolversForResolveType: false,
		},
	}),
);
