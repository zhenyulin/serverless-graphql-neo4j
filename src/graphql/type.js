export default `
	type User {
		id: ID!
		name: String
		email: String!
		follows: [User] @relation(name: "FOLLOWS", direction: "OUT")
		followers: [User] @relation(name: "FOLLOWS", direction: "IN")
		likes: [Item] @relation(name: "LIKES", direction: "OUT")
	}

	type Item {
		id: ID!
		name: String
		likedBy: [User] @relation(name: "LIKES", direction: "IN")
	}

	type Query {
		user(id: ID!): User
		item(id: ID!): Item
	}
`;
