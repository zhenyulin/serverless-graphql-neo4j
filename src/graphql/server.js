import express from 'express';
import serverless from 'serverless-http';
import expressGraphql from 'express-graphql';
import { makeExecutableSchema } from 'graphql-tools';

import typeDefs from './type';
import resolvers from './resolver';

const schema = makeExecutableSchema({ typeDefs, resolvers });
const handler = expressGraphql({
	schema,
	graphiql: true,
});
const app = express().use(handler);

export default serverless(app);
