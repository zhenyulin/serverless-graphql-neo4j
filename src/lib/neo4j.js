import { v1 as neo4j } from 'neo4j-driver';

const driver = neo4j.driver(
	process.env.NEO4J_URI,
	neo4j.auth.basic(process.env.NEO4J_USER, process.env.NEO4J_PASSWORD),
);

export const runCypher = async (cypher, param, debug) => {
	if (debug) console.log(cypher);
	const session = driver.session();
	try {
		const { records } = await session.run(cypher, param);
		const items = records.map(r => r.get(0).properties);
		if (debug) console.log(items);
		return items;
	} catch (e) {
		console.log(e);
		throw e;
	} finally {
		session.close();
	}
};

export const runCypherReturnOne = async (cypher, param, debug) => {
	if (debug) console.log(cypher);
	const session = driver.session();
	try {
		const { records } = await session.run(cypher, param);
		const items = records.map(r => r.get(0).properties);
		if (debug) console.log(items);
		return items[0];
	} catch (e) {
		if (
			e.code &&
			e.code === 'Neo.ClientError.Schema.ConstraintValidationFailed'
		) {
			e.message = 'Node with constraint properties already exists';
		}
		throw e;
	} finally {
		session.close();
	}
};

export const runTransactionWithArray = async (cypher, params, debug) => {
	const session = driver.session();
	const transaction = session.beginTransaction();
	try {
		const result = [];
		params.forEach(async param => {
			if (debug) console.log(cypher, param);
			const r = await transaction.run(cypher, param);
			result.push(r);
		});
		await transaction.commit();
		return result;
	} catch (e) {
		transaction.rollback();
		throw e;
	} finally {
		session.close();
	}
};

export const runCyphersAsTransactions = async cyphers => {
	const session = driver.session();
	const transaction = session.beginTransaction();
	try {
		const result = [];
		cyphers.forEach(async cypher => {
			const r = await transaction.run(cypher);
			result.push(r);
		});
		await transaction.commit();
		return result;
	} catch (e) {
		transaction.rollback();
		throw e;
	} finally {
		session.close();
	}
};

export default driver;
