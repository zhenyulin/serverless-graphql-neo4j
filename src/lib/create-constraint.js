import { runCypher } from './neo4j';

export const createUserConstraint = () =>
	runCypher(`
	CREATE CONSTRAINT ON (u:User) ASSERT u.email IS UNIQUE
`);

export const createItemConstraint = () =>
	runCypher(`
	CREATE CONSTRAINT ON (i:Item) ASSERT i.name IS UNIQUE
`);

export default async () => {
	await Promise.all([createUserConstraint(), createItemConstraint()]);
};
