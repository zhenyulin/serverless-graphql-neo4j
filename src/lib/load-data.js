import * as Fixtures from '__fixtures__';

import { runCypher, runTransaction } from './neo4j';

export const createUsers = () =>
	runTransaction(
		`CREATE (u:User $user) RETURN u`,
		Fixtures.USERS.map(user => ({ user })),
	);

export const createItems = () =>
	runTransaction(
		`CREATE (i:Item $item) RETURN i`,
		Fixtures.ITEMS.map(item => ({ item })),
	);

export const createUserFollowUsers = () =>
	runTransaction(
		`MATCH (follower:User), (followee:User)
			WHERE follower.id = $followerId AND followee.id = $followeeId
			CREATE (follower)-[r:FOLLOW]->(followee)
			RETURN r`,
		Fixtures.USER_FOLLOW_USERS,
	);

export const createUserLikeItems = () =>
	runTransaction(
		`MATCH (user:User), (item:Item)
			WHERE user.id = $userId AND item.id = $itemId
			CREATE (user)-[r:LIKE]->(item)
			RETURN r`,
		Fixtures.USER_LIKE_ITEMS,
	);

export const loadData = async () => {
	await createUsers();
	await createItems();
	await createUserFollowUsers();
	await createUserLikeItems();
};

export const emptyData = () => runCypher(`MATCH (n) DETACH DELETE n`);
