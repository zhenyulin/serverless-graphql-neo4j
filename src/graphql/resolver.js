import { neo4jgraphql } from 'neo4j-graphql-js';
import { runCypher } from 'lib/neo4j';

export default {
	Query: {
		User: (...args) => neo4jgraphql(...args, false),
		Users: (...args) => neo4jgraphql(...args, false),
		Item: (...args) => neo4jgraphql(...args, false),
		Items: (...args) => neo4jgraphql(...args, false),
	},
	User: {
		followees: ({ id }) =>
			runCypher(
				`
			MATCH (:User{id:${id}})-[:FOLLOW]->(followees:User)
			RETURN followees
		`,
				{},
				true,
			),
		followers: ({ id }) =>
			runCypher(
				`
			MATCH (:User{id:${id}})<-[:FOLLOW]->(followers:User)
			RETURN followers
		`,
				{},
				true,
			),
	},
	Mutation: {
		UserFollowUser: (_, { followerId, followeeId }) =>
			runCypher(`
				MATCH (follower:User {id: "${followerId}"})
				MATCH (followee:User {id: "${followeeId}"})
				MERGE (follower)-[:FOLLOW]->(followee)
				RETURN follower
			`),
		UserUnfollowUser: async (_, { followerId, followeeId }) =>
			runCypher(`
					MATCH (follower:User {id: "${followerId}"})
					MATCH (followee:User {id: "${followeeId}"})
					OPTIONAL MATCH (follower)-[follow:FOLLOW]->(followee)
					DELETE follow
					RETURN follower
				`),
		UserLikeItem: async (_, { userId, itemId }) =>
			runCypher(`
					MATCH (user:User {id: "${userId}"})
					MATCH (item:Item {id: "${itemId}"})
					MERGE (user)-[:LIKE]->(item)
					RETURN user { .*, likeItems: [(user)-[:LIKE]->(items:Item) | items { .* }] } AS user
				`),
		UserDislikeItem: async (_, { userId, itemId }) =>
			runCypher(`
					MATCH (user:User {id: "${userId}"})
					MATCH (item:Item {id: "${itemId}"})
					OPTIONAL MATCH (user)-[like:LIKE]->(item)
					DELETE like
					RETURN user { .*, likeItems: [(user)-[:LIKE]->(items:Item) | items { .* }] } AS user
				`),
	},
};
