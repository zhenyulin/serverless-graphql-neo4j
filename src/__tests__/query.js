import { graphql } from 'graphql';

import driver from 'lib/neo4j';
import { emptyData, loadData } from 'lib/load-data';
import schema from 'graphql/schema';

const testCases = cases =>
	Object.keys(cases).map(caseName =>
		it(caseName, async () => {
			const query = cases[caseName];
			const result = await graphql(schema, query);
			expect(result).toMatchSnapshot();
		}),
	);

describe('Query', () => {
	beforeAll(async () => {
		await emptyData();
		await loadData();
	});

	afterAll(() => {
		driver.close();
	});

	describe('User', () => {
		testCases({
			'get user by id': `
				query {
					User(id: "4edd40c86762e0fb12000003") {
						id
						name
						followees {
							name
						}
						followers {
							name
						}
						likeItems {
							name
						}
					}
				}
			`,
			'get similarUserLiked': `
				query {
					User(id: "4edd40c86762e0fb12000004") {
						similarUsersLiked {
							name
						}
					}
				}
			`,
			'return null if user not found': `
				query {
					User(id: "4edd40c86762e0fb12000099") {
						id
						name
					}
				}
			`,
			'throw error if id is not provided': `
				query {
					User() {
						id
						name
					}
				}
			`,
			'throw error if id is not valid id': `
				query {
					User(id: "random") {
						id
						name
					}
				}
			`,
		});
	});

	describe('Users', () => {
		testCases({
			'get all users': `
				query {
					Users {
						id
						name
						followees {
							name
						}
						followers {
							name
						}
						likeItems {
							name
						}
					}
				}
			`,
		});
	});

	describe('Item', () => {
		testCases({
			'get item by id': `
				query {
					Item(id: "4edd40c86762e0fb12000015") {
						id
						name
						likedByUsers {
							name
						}
					}
				}
			`,
			'get item.ratedByUsers': `
				query {
					Item(id: "4edd40c86762e0fb12000014") {
						ratedByUsers {
							user {
								name
							}
							rating
						}
					}
				}
			`,
			'get item.averageRating': `
				query {
					Item(id: "4edd40c86762e0fb12000014") {
						averageRating
					}
				}
			`,
			'return null if item not found': `
				query {
					Item(id: "4edd40c86762e0fb12000099") {
						id
						name
					}
				}
			`,
			'throw error if id is not provided': `
				query {
					Item() {
						id
						name
					}
				}
			`,
			'throw error if id is not valid id': `
				query {
					Item(id: "random") {
						id
						name
					}
				}
			`,
		});
	});

	describe('Items', () => {
		testCases({
			'get all items': `
				query {
					Items {
						id
						name
						likedByUsers {
							name
						}
					}
				}
			`,
		});
	});
});
