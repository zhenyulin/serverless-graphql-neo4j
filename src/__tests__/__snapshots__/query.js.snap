// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Item get item by id 1`] = `
Object {
  "data": Object {
    "Item": Object {
      "id": "4edd40c86762e0fb12000015",
      "likedByUsers": Array [],
      "name": "Birth of Cool",
    },
  },
}
`;

exports[`Query Item return null if item not found 1`] = `
Object {
  "data": Object {
    "Item": null,
  },
}
`;

exports[`Query Item throw error if id is not provided 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Syntax Error: Expected Name, found )],
  ],
}
`;

exports[`Query Item throw error if id is not valid id 1`] = `
Object {
  "data": Object {
    "Item": null,
  },
}
`;

exports[`Query Items get all items 1`] = `
Object {
  "data": Object {
    "Items": Array [
      Object {
        "id": "4edd40c86762e0fb12000013",
        "likedByUsers": Array [],
        "name": "Imagine",
      },
      Object {
        "id": "4edd40c86762e0fb12000015",
        "likedByUsers": Array [],
        "name": "Birth of Cool",
      },
      Object {
        "id": "4edd40c86762e0fb12000014",
        "likedByUsers": Array [
          Object {
            "name": "John Lennon",
          },
        ],
        "name": "4:00",
      },
    ],
  },
}
`;

exports[`Query User get user by id 1`] = `
Object {
  "data": Object {
    "User": Object {
      "followees": Array [],
      "followers": Array [
        Object {
          "name": "John Lennon",
        },
      ],
      "id": "4edd40c86762e0fb12000003",
      "likeItems": Array [],
      "name": "John Cage",
    },
  },
}
`;

exports[`Query User return null if user not found 1`] = `
Object {
  "data": Object {
    "User": null,
  },
}
`;

exports[`Query User throw error if id is not provided 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Syntax Error: Expected Name, found )],
  ],
}
`;

exports[`Query User throw error if id is not valid id 1`] = `
Object {
  "data": Object {
    "User": null,
  },
}
`;

exports[`Query Users get all users 1`] = `
Object {
  "data": Object {
    "Users": Array [
      Object {
        "followees": Array [],
        "followers": Array [],
        "id": "4edd40c86762e0fb12000005",
        "likeItems": Array [],
        "name": "Miles Davis",
      },
      Object {
        "followees": Array [
          Object {
            "name": "John Cage",
          },
        ],
        "followers": Array [],
        "id": "4edd40c86762e0fb12000004",
        "likeItems": Array [
          Object {
            "name": "4:00",
          },
        ],
        "name": "John Lennon",
      },
      Object {
        "followees": Array [],
        "followers": Array [
          Object {
            "name": "John Lennon",
          },
        ],
        "id": "4edd40c86762e0fb12000003",
        "likeItems": Array [],
        "name": "John Cage",
      },
    ],
  },
}
`;
